flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -236,6 +238,9 @@ private void setTables(ContextResolvedTable catalogTable) {
                 tables.add(catalogTable);
             } else {
                 for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
+                    if (tables.contains(catalogTable)) {
```

Comment:
I think we can use a boolean flag to check here, then we don't need to call `contains` method every time, it is O(N) time complexity.

```
                boolean hasAdded = false;
                for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
                    if (hasAdded) {
                        break;
                    }
                    if (!thisTable.getIdentifier().equals(catalogTable.getIdentifier())) {
                        tables.add(catalogTable);
                        hasAdded = true;
                    }
                }
```

Reply:
> 我想我们可以在这里使用布尔标志来检查，那么我们不需要每次都调用方法，它是 O（N） 时间复杂度。`contains`
> 
> ```
>                 boolean hasAdded = false;
>                 for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
>                     if (hasAdded) {
>                         break;
>                     }
>                     if (!thisTable.getIdentifier().equals(catalogTable.getIdentifier())) {
>                         tables.add(catalogTable);
>                         hasAdded = true;
>                     }
>                 }
> ```

I think we should modify the traversal logic.
```
                boolean hasAdded = false;
                for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
                    if (thisTable.getIdentifier().equals(catalogTable.getIdentifier())) {
                        hasAdded = true;
                        break;
                    }
                }
                if (!hasAdded) {
                    tables.add(catalogTable);
                }
```

Reply:
HI @mumuhhh thanks for the ping. I think you are right. Now that I look in more detail, in addition to your suggestion, I think we can also remove the first  `if` check in the method.